### PR TITLE
[NT-1963] Show removed comments

### PR DESCRIPTION
**Pull request by singhhari**
# 📲 What

This PR adds some functionality to our `CommentRemovedCell` view. In particular, we're enabling the view of replies even though a comment may be flagged. Moreover, we're rendering a web view linked to our Community Guidelines when the attributed text on the cell is tapped.

# 🤔 Why

We're aiming for parity with our website when we include features like this. Not only does the website allow a user to view replies on a flagged comment but it also links a user to [Community Guidelines](https://www.kickstarter.com/help/community) through a hyperlink in the cell.

# 🛠 How

The existing implementation on the `CommentRemovedCell` was fairly barebones. We re-factored the label that was rendering the disclaimer text to a `UITextView` so we could leverage the `UITextViewDelegate` methods that can detect when an URL is selected. 

# 👀 See

https://user-images.githubusercontent.com/15615702/121608997-f0a45d80-ca20-11eb-888a-b94a6bce38a8.mp4

# ✅ Acceptance criteria

- [x] Navigate to a project with flagged comments like [Kurv Guitar](https://www.kickstarter.com/projects/kurv/kurv-guitar) and verify the attributed text opens up the Community Guidelines webpage upon selection.
- [x] Also navigate to a project like [Shenmue 3](https://www.kickstarter.com/projects/ysnet/shenmue-3) and verify a flagged comment with replies on it like [this one](https://www.kickstarter.com/projects/ysnet/shenmue-3/comments?comment=Q29tbWVudC0zMTc5OTczMg%3D%3D) is rendering appropriately.

### CODE REVIEW

File: Library/HelpType.swift [hunk 1]
```diff
@@ -23,6 +24,8 @@ public enum HelpType: SettingsCellTypeProtocol, CaseIterable {
     switch self {
     case .helpCenter:
       return Strings.Help_center()
+    case .community:
+      return ""
```
Comment (justinswart) [inline Library/HelpType.swift:28]: TBD?

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift [hunk 2]
```diff
@@ -87,6 +103,21 @@ final class CommentRemovedCell: UITableViewCell, ValueCell {
       self.infoImageView.heightAnchor.constraint(equalToConstant: Styles.grid(4))
     ])
   }
+
+  // MARK: - View model
+
+  internal override func bindViewModel() {
+    super.bindViewModel()
+
+    self.viewRepliesView.rac.hidden = self.viewModel.outputs.viewRepliesStackViewIsHidden
```
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:112]: Let's also rename the output -> `viewRepliesViewHidden`.

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift [hunk 3]
```diff
@@ -99,27 +130,64 @@ private let rowStackViewStyle: StackViewStyle = { stackView in
...
     attributes: regularFontAttribute
   )
-  let learnMoreAttributedString = NSMutableAttributedString(
-    string: Strings.Learn_more_about_comment_guidelines(),
-    attributes: coloredFontAttribute
-  )
-  attributedString.append(learnMoreAttributedString)
 
-  return attributedString
+  guard let communityGuidelinesLink = HelpType.community
```
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:149]: Wondering if it makes sense to hardcode `community` in two places or if the delegate should pass back the `HelpType` 🤔

File: Library/ViewModels/CommentsViewModel.swift [hunk 4]
```diff
@@ -257,6 +263,8 @@ public final class CommentsViewModel: CommentsViewModelType,
       errors.mapConst(.error)
     )
     .skipRepeats()
+
+    self.showHelpWebViewController = self.commentRemovedCellDidTapURLProperty.signal.mapConst(.community)
```
Comment (justinswart) [inline Library/ViewModels/CommentsViewModel.swift:267]: Referring to this other hard-coding here.

**Review by justinswart — APPROVED** [review]
Cool looks good, small suggestions.

File: Library/HelpType.swift (hunk 1 above)
Comment (singhhari) [reply Library/HelpType.swift:28]: So we're doing the same thing for the `trust` enum case here. Not sure If I should leave this blank as well or submit a new string like "Community Guidelines" for translation?

**Review by msadoon — APPROVED** [review]
Good to go!

One issue I found on iPad was the border for view replies not appearing. It does appear for iPhone.

File: Kickstarter-iOS/Views/Controllers/CommentsViewControllerTests.swift [hunk 5]
```diff
@@ -41,6 +41,38 @@ internal final class CommentsViewControllerTests: TestCase {
     }
   }
 
+  func testView_WithFailedRemovedAndSuccessfulComments_ShouldDisplayAll_CommentThreadingRepliesEnabledFeatureFlag_True() {
```
[the remark this answers is not in the document]
Comment (msadoon) [reply Kickstarter-iOS/Views/Controllers/CommentsViewControllerTests.swift:44]: I mean the test name above this one, for some reason GH doesn't allow commenting on lines of code that were not changed.

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift (hunk 3 above)
Comment (singhhari) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:149]: Good catch. Removing this line here in favor of the `mapConst` in `CommentsViewModel`.

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift [hunk 6]
```diff
@@ -99,27 +130,64 @@ private let rowStackViewStyle: StackViewStyle = { stackView in
...
+    return removedCommentAttributedString
+  }
+
+  let communityGuidelinesString = Strings
+    .Learn_more_about_comment_guidelines(community_link: communityGuidelinesLink)
+
+  guard let communityGuidelinesAttributedString = try? NSMutableAttributedString(
+    data: Data(communityGuidelinesString.utf8),
+    options: [
+      .documentType: NSAttributedString.DocumentType.html,
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:160]: Yup, as long as you wrap the text in HTML the textViewDelegate method will interpret it as such.

File: Kickstarter.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -198,6 +198,7 @@
 		47AC212F264AD2590090AEDF /* CommentsViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 47AC212E264AD2590090AEDF /* CommentsViewController.swift */; };
 		47AC2139264AD5430090AEDF /* CommentsViewModelTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 47AC2138264AD5430090AEDF /* CommentsViewModelTests.swift */; };
 		47AC213E264AD7E00090AEDF /* CommentsViewControllerTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 47AC213D264AD7E00090AEDF /* CommentsViewControllerTests.swift */; };
+		47F95ED72672C594001365B2 /* ViewRepliesView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 47F95ECF2672C587001365B2 /* ViewRepliesView.swift */; };
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Kickstarter.xcodeproj/project.pbxproj:201]: Good catch. Updated.

File: Library/ViewModels/CommentCellViewModelTests.swift [hunk 8]
```diff
@@ -150,6 +152,30 @@ internal final class CommentCellViewModelTests: TestCase {
     }
   }
 
+  func testOutput_notifyDelegateLinkTappedWithURL() {
+    let mockOptimizelyClient = MockOptimizelyClient()
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Library/ViewModels/CommentCellViewModelTests.swift:156]: Yup, the flag is defaulting to false which means the view will not be visible.

[the remark this answers is not in the document]
Comment (singhhari) [reply Library/ViewModels/CommentCellViewModelTests.swift:156]: Also btw a very annoying thing we will have to clean up when the flag is deprecated...

File: Library/ViewModels/CommentsViewModelTests.swift [hunk 9]
```diff
@@ -137,6 +139,22 @@ internal final class CommentsViewModelTests: TestCase {
     }
   }
 
+  func testOutput_ShowHelpWebViewController() {
+    self.showHelpWebViewController.assertDidNotEmitValue()
+
+    withEnvironment {
+      self.vm.inputs.configureWith(project: .template, update: nil)
+      self.vm.inputs.viewDidLoad()
+
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Library/ViewModels/CommentsViewModelTests.swift:148]: Makes sense. Updated.

File: Kickstarter-iOS/Views/Controllers/CommentsViewControllerTests.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (singhhari) [reply Kickstarter-iOS/Views/Controllers/CommentsViewControllerTests.swift:44]: Ah good catch. Updated.

File: Kickstarter-iOS/Views/Controllers/CommentsViewControllerTests.swift [hunk 10]
```diff
@@ -41,6 +41,38 @@ internal final class CommentsViewControllerTests: TestCase {
...
+      withEnvironment(
+        apiService: mockService,
+        currentUser: .template,
+        language: language,
+        optimizelyClient: mockOptimizelyClient
+      ) {
+        let controller = CommentsViewController.configuredWith(project: Project.template)
+
+        let (parent, _) = traitControllers(
+          device: .phone4_7inch,
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Kickstarter-iOS/Views/Controllers/CommentsViewControllerTests.swift:62]: Seems like we forgot to add the `combo` closure that wraps each test with our devices. Just added it into all the tests in here.

**Comment by singhhari** [comment]
> Good to go!
> 
> One issue I found on iPad was the border for view replies not appearing. It does appear for iPhone.
> 
> 

Pinged you about this Slack but for record-keeping purposes I am seeing this on my iPad simulator so I think that may have been a minor hiccup on the sim @msadoon tested on.

**Review by moderateepheezy — APPROVED** [review]

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift (hunk 3 above)
Comment (justinswart) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift]: I guess what I'm saying is that maybe we should just pass the URL through? The hard-coding prevents us from easily using a different URL in the tappable text or maybe more than one tappable link for example.

File: Library/HelpType.swift (hunk 1 above)
Comment (justinswart) [reply Library/HelpType.swift:28]: Oh I guess these titles are for when we used to show the activity sheet, I don't think we do anymore so we can leave this for now.

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift (hunk 3 above)
Comment (justinswart) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift]: We do this here: https://github.com/kickstarter/ios-oss/blob/master/Library/ViewModels/PledgeViewCTAContainerViewModel.swift#L44

Maybe we should make that a static func on `HelpType` for reusability.

Comment (singhhari) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:149]: Updated!

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**